nf_core/download.py

```diff
@@ -664,13 +664,53 @@ def find_container_images(self, workflow_directory):
         """Find container image names for workflow.
 
         Starts by using `nextflow config` to pull out any process.container
-        declarations. This works for DSL1. It should return a simple string with resolved logic.
+        declarations. This works for DSL1. It should return a simple string with resolved logic,
+        but not always, e.g. not for differentialabundance 1.2.0
 
         Second, we look for DSL2 containers. These can't be found with
         `nextflow config` at the time of writing, so we scrape the pipeline files.
-        This returns raw source code that will likely need to be cleaned.
+        This returns raw matches that will likely need to be cleaned.
+        """
+
+        log.debug("Fetching container names for workflow")
+        # since this is run for multiple revisions now, account for previously detected containers.
+        previous_findings = [] if not self.containers else self.containers
+        config_findings = []
+        module_findings = []
+
+        # Use linting code to parse the pipeline nextflow config
+        self.nf_config = nf_core.utils.fetch_wf_config(workflow_directory)
+
+        # Find any config variables that look like a container
+        for k, v in self.nf_config.items():
+            if k.startswith("process.") and k.endswith(".container"):
+                config_findings.append(v.strip('"').strip("'"))
+
+        # rectify the container paths found in the config
+        # Raw config_findings may yield multiple containers, so better create a shallow copy of the list, since length of input and output may be different ?!?
+        config_findings = self.rectify_raw_container_matches(config_findings[:], self.nf_config, "modules.config")
+
+        # Recursive search through any DSL2 module files for container spec lines.
+        for subdir, _, files in os.walk(os.path.join(workflow_directory, "modules")):
+            for file in files:
+                if file.endswith(".nf"):
+                    file_path = os.path.join(subdir, file)
+                    with open(file_path, "r") as fh:
+                        # Look for any lines with `container = "xxx"`
+                        search_space = fh.read()
+                        module_container = re.findall(r"container\s*[\"\']([^\"]*)[\"\']", search_space, re.S)
```

Comment:
Would adding the newline solve #2338 ?
```suggestion
                        module_container = re.findall(r"container\s*[\"\']([^\"]*)[\"\']\n", search_space, re.S)
```

Reply:
Sadly, not because there are modules where the DSL2 is wrapped into curly braces, so the quote is followed by a closing bracket. There are also modules with escaped quotes inside unescaped quotes that need to be matched.

After many hours, I realized that capturing the quote and using exactly this instead of an either or notation `[\"\']` is the only thing that prevented abridged matches.
```
 container_regex = re.compile(
                            r"container\s+[\s{}=$]*(?P<quote>[\'\"])(?P<param>(?:.(?!\1))*.?)\1[\s}]*",
                            re.DOTALL
                        )

local_module_findings = re.findall(container_regex, search_space)
```

I have now tested about 20+ pipelines in multiple revisions, and it seems that will work for all modules and configs in there. There are still some containers not picked up (e.g. in Sarek 2.5), but this is not due to the regex, but due to the downstream logic how to prioritize https:// direct download container addresses over those that would need to be pulled with Singularity.

_Maybe we should in the long-term either have a nf-core container API or include a small sqlite-Database with curated container addresses for all revisions into tools?_